dandiapi/api/tests/test_unembargo.py

```diff
@@ -0,0 +1,230 @@
+import math
+import os
+
+from dandischema.digests.dandietag import PartGenerator, mb
+from django.contrib.auth.models import AnonymousUser
+from django.core.files.uploadedfile import SimpleUploadedFile
+import factory
+from guardian.shortcuts import assign_perm
+import pytest
+
+from dandiapi.api import tasks
+from dandiapi.api.models import Asset, AssetBlob, EmbargoedAssetBlob, Version
+from dandiapi.api.models.dandiset import Dandiset
+
+
+@pytest.mark.parametrize(
+    ('embargo_status', 'user_status', 'resp_code'),
+    [
+        (Dandiset.EmbargoStatus.OPEN, 'owner', 400),
+        (Dandiset.EmbargoStatus.OPEN, 'anonymous', 401),
+        (Dandiset.EmbargoStatus.OPEN, 'not-owner', 403),
+        (Dandiset.EmbargoStatus.EMBARGOED, 'owner', 200),
+        (Dandiset.EmbargoStatus.EMBARGOED, 'anonymous', 401),
+        (Dandiset.EmbargoStatus.EMBARGOED, 'not-owner', 403),
+        (Dandiset.EmbargoStatus.UNEMBARGOING, 'owner', 400),
+        (Dandiset.EmbargoStatus.UNEMBARGOING, 'anonymous', 401),
+        (Dandiset.EmbargoStatus.UNEMBARGOING, 'not-owner', 403),
+    ],
+)
+@pytest.mark.django_db
+def test_dandiset_rest_unembargo(
+    api_client, dandiset_factory, user_factory, embargo_status, user_status, resp_code
+):
+    dandiset: Dandiset = dandiset_factory(embargo_status=embargo_status)
+    if user_status == 'anonymous':
+        user = AnonymousUser
+    else:
+        user = user_factory()
+        api_client.force_authenticate(user=user)
+    if user_status == 'owner':
+        assign_perm('owner', user, dandiset)
+
+    response = api_client.post(f'/api/dandisets/{dandiset.identifier}/unembargo/')
+    assert response.status_code == resp_code
+
+
+@pytest.mark.parametrize(
+    ('file_size', 'part_size'),
+    [
+        (100, mb(64)),  # Normal
+        (mb(30), mb(5)),  # Few parts
+        (mb(100), mb(6)),  # Many parts (tests concurrency)
+    ],
+)
+@pytest.mark.django_db
+def test_unembargo_dandiset(
+    dandiset_factory,
+    draft_version_factory,
+    asset_factory,
+    embargoed_asset_blob_factory,
+    storage,
+    embargoed_storage,
+    file_size,
+    part_size,
+    monkeypatch,
+):
+    # Since both fixtures are parametrized, only proceed when they use the same storage backend
+    if type(storage) != type(embargoed_storage):
+        pytest.skip('Skip tests with mismatched storages')
+
+    # Pretend like AssetBlob/EmbargoedAssetBlob were defined with the given storage
+    monkeypatch.setattr(AssetBlob.blob.field, 'storage', storage)
+    monkeypatch.setattr(EmbargoedAssetBlob.blob.field, 'storage', embargoed_storage)
```

Comment:
this looks much more sophisticated than what I've been doing. Does it reset the values after the test completes?

Reply:
I believe so https://docs.pytest.org/en/6.2.x/monkeypatch.html

Namely

> All modifications will be undone after the requesting test function or fixture has finished. The raising parameter determines if a KeyError or AttributeError will be raised if the target of the set/deletion operation does not exist.

Which I really like :)